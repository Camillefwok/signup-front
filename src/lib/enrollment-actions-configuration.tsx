import { ReactElement } from 'react';
import DoneIcon from '../components/atoms/icons/done';

export enum EnrollmentAction {
  notify,
  destroy,
  update,
  send_application,
  refuse_application,
  review_application,
  validate_application,
}

export type ActionConfiguration = {
  label: string;
  cssClass: string;
  icon?: ReactElement;
  needsToComputeNextEnrollmentState?: boolean;
  promptForComment?: boolean;
};

export const userInteractionsConfiguration: {
  [key in keyof typeof EnrollmentAction]: ActionConfiguration
} = {
  notify: {
    label: 'Envoyer un message',
    cssClass: 'secondary',
    needsToComputeNextEnrollmentState: true,
    promptForComment: true,
  },
  destroy: {
    label: 'Supprimer la demande',
    cssClass: 'warning',
  },
  update: {
    label: 'Sauvegarder le brouillon',
    cssClass: 'secondary',
  },
  send_application: {
    label: 'Soumettre la demande',
    icon: <DoneIcon color="white" />,
    cssClass: 'primary',
    needsToComputeNextEnrollmentState: true,
  },
  refuse_application: {
    label: 'Refuser',
    cssClass: 'warning',
    needsToComputeNextEnrollmentState: true,
    promptForComment: true,
  },
  review_application: {
    label: 'Demander une modification',
    cssClass: 'secondary',
    needsToComputeNextEnrollmentState: true,
    promptForComment: true,
  },
  validate_application: {
    label: 'Valider',
    cssClass: 'primary',
    needsToComputeNextEnrollmentState: true,
    promptForComment: true,
  },
};
