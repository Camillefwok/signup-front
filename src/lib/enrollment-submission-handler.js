import { getErrorMessages } from '.';
import {
  createOrUpdateEnrollment,
  deleteEnrollment,
  computeNextEnrollmentState,
} from '../services/enrollments';

export const handleEnrollmentSubmission = async (
  actionConfiguration,
  setIntendedAction,
  enrollment,
  waitForUserInteractionInPrompt,
  updateEnrollment
) => {
  const resultMessages = {
    errorMessages: [],
    successMessages: [],
    redirectToHome: false,
  };

  const action = actionConfiguration.id;

  try {
    let comment = null;
    let commentFullEditMode = null;

    if (actionConfiguration.promptForComment) {
      try {
        setIntendedAction(action);

        const actionMessage = await waitForUserInteractionInPrompt;
        comment = actionMessage.message;
        commentFullEditMode = actionMessage.fullEditMode;
      } catch (e) {
        return resultMessages;
      }
    }

    let enrollmentId = enrollment.id;

    if (enrollment.acl.update) {
      const newEnrollment = await createOrUpdateEnrollment({
        enrollment,
      });
      updateEnrollment(newEnrollment);
      enrollmentId = newEnrollment.id;

      resultMessages.successMessages.push('Votre demande a été sauvegardée.');
    }

    if (action === 'update') {
      return resultMessages;
    }

    if (action === 'destroy') {
      await deleteEnrollment({ id: enrollmentId });
    }

    if (actionConfiguration.needsToComputeNextEnrollmentState) {
      await computeNextEnrollmentState({
        action,
        id: enrollmentId,
        comment,
        commentFullEditMode,
      });
    }

    resultMessages.redirectToHome = true;

    return resultMessages;
  } catch (error) {
    resultMessages.errorMessages.push(...getErrorMessages(error));

    return resultMessages;
  }
};
