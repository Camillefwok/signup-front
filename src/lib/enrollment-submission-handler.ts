import { getErrorMessages } from '.';
import {
  ActionConfiguration,
  EnrollmentAction,
} from './enrollment-actions-configuration';
import {
  createOrUpdateEnrollment,
  deleteEnrollment,
  computeNextEnrollmentState,
} from '../services/enrollments';

export const croute = async (
  action: EnrollmentAction,
  actionConfiguration: ActionConfiguration,
  enrollment: any,
  updateEnrollment: Function,
  message?: string,
  fullEditMode?: boolean
): Promise<{
  redirectToHome: boolean;
  successMessages: string[];
  errorMessages: string[];
}> => {
  const resultMessages = {
    errorMessages: [] as string[],
    successMessages: [] as string[],
    redirectToHome: false,
  };

  try {
    let comment = null;
    let commentFullEditMode = null;

    if (actionConfiguration.promptForComment) {
      try {
        comment = message;
        commentFullEditMode = fullEditMode;
      } catch (e) {
        return resultMessages;
      }
    }

    let enrollmentId = enrollment.id;

    if (enrollment.acl.update) {
      const newEnrollment = await createOrUpdateEnrollment({
        enrollment,
      });
      updateEnrollment(newEnrollment);
      enrollmentId = newEnrollment.id;

      resultMessages.successMessages.push('Votre demande a été sauvegardée.');
    }

    if (action === 'update') {
      return resultMessages;
    }

    if (action === 'destroy') {
      await deleteEnrollment({ id: enrollmentId });
    }

    if (actionConfiguration.needsToComputeNextEnrollmentState) {
      await computeNextEnrollmentState({
        action,
        id: enrollmentId,
        comment,
        commentFullEditMode,
      });
    }

    resultMessages.redirectToHome = true;

    return resultMessages;
  } catch (error) {
    resultMessages.errorMessages.push(...getErrorMessages(error));

    return resultMessages;
  }
};

export const handleEnrollmentSubmission = async (
  actionConfiguration: ActionConfiguration & { id: EnrollmentAction },
  setIntendedAction: Function,
  enrollment: any,
  waitForUserInteractionInPrompt: Promise<{
    message: string;
    fullEditMode: boolean;
  }>,
  updateEnrollment: Function
) => {
  const resultMessages = {
    errorMessages: [] as string[],
    successMessages: [] as string[],
    redirectToHome: false,
  };

  const action = actionConfiguration.id;

  try {
    let comment = null;
    let commentFullEditMode = null;

    if (actionConfiguration.promptForComment) {
      try {
        setIntendedAction(action);

        const actionMessage = await waitForUserInteractionInPrompt;
        comment = actionMessage.message;
        commentFullEditMode = actionMessage.fullEditMode;
      } catch (e) {
        return resultMessages;
      }
    }

    let enrollmentId = enrollment.id;

    if (enrollment.acl.update) {
      const newEnrollment = await createOrUpdateEnrollment({
        enrollment,
      });
      updateEnrollment(newEnrollment);
      enrollmentId = newEnrollment.id;

      resultMessages.successMessages.push('Votre demande a été sauvegardée.');
    }

    if (action === 'update') {
      return resultMessages;
    }

    if (action === 'destroy') {
      await deleteEnrollment({ id: enrollmentId });
    }

    if (actionConfiguration.needsToComputeNextEnrollmentState) {
      await computeNextEnrollmentState({
        action,
        id: enrollmentId,
        comment,
        commentFullEditMode,
      });
    }

    resultMessages.redirectToHome = true;

    return resultMessages;
  } catch (error) {
    resultMessages.errorMessages.push(...getErrorMessages(error));

    return resultMessages;
  }
};
