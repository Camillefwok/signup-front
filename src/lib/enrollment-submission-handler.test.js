import { userInteractionsConfiguration } from './enrollment-actions-configuration';
jest.mock('../services/enrollments');
// eslint-disable-next-line import/first
import { handleEnrollmentSubmission } from './enrollment-submission-handler';
// eslint-disable-next-line import/first
import {
  computeNextEnrollmentState,
  deleteEnrollment,
  createOrUpdateEnrollment,
} from '../services/enrollments';

describe('When submitting the enrollment form', () => {
  let userCancelsPrompt;
  let userValidatesPrompt;
  let waitForUserInteractionInPrompt;

  const setIntendedAction = jest.fn();
  const enrollment = { id: Symbol(), acl: {} };
  const updateEnrollment = jest.fn();

  beforeEach(() => {
    jest.resetAllMocks();
    waitForUserInteractionInPrompt = new Promise((resolve, reject) => {
      userValidatesPrompt = resolve;
      userCancelsPrompt = reject;
    });
  });

  describe('with the notify action', () => {
    const actionConfiguration = {
      id: 'notify',
      ...userInteractionsConfiguration.notify,
    };

    it('waits for user prompt', async () => {
      userCancelsPrompt();
      const output = await handleEnrollmentSubmission(
        actionConfiguration,
        setIntendedAction,
        enrollment,
        waitForUserInteractionInPrompt,
        updateEnrollment
      );

      expect(setIntendedAction).toHaveBeenCalledWith('notify');
      expect(computeNextEnrollmentState).not.toHaveBeenCalled();
      expect(output).toMatchSnapshot();
    });

    it('calls for the enrollment state update after user validates prompt', async () => {
      const userMessage = 'La barbe de la femme à Georges Moustaki';

      userValidatesPrompt({ message: userMessage, fullEditMode: false });
      const output = await handleEnrollmentSubmission(
        actionConfiguration,
        setIntendedAction,
        enrollment,
        waitForUserInteractionInPrompt,
        updateEnrollment
      );

      expect(computeNextEnrollmentState).toHaveBeenCalledWith({
        action: 'notify',
        comment: userMessage,
        commentFullEditMode: false,
        id: enrollment.id,
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('with the destroy action', () => {
    const actionConfiguration = {
      id: 'destroy',
      ...userInteractionsConfiguration.destroy,
    };

    it('calls the delete endpoint', async () => {
      const output = await handleEnrollmentSubmission(
        actionConfiguration,
        setIntendedAction,
        enrollment,
        waitForUserInteractionInPrompt,
        updateEnrollment
      );

      expect(deleteEnrollment).toHaveBeenCalledWith({
        id: enrollment.id,
      });
      expect(output).toMatchSnapshot();
    });
  });

  describe('with the update action', () => {
    const actionConfiguration = {
      id: 'update',
      ...userInteractionsConfiguration.update,
    };
    const enrollmentToUpdate = { ...enrollment, acl: { update: true } };

    it('calls the update endpoint', async () => {
      createOrUpdateEnrollment.mockResolvedValue(enrollmentToUpdate);

      const output = await handleEnrollmentSubmission(
        actionConfiguration,
        setIntendedAction,
        enrollmentToUpdate,
        waitForUserInteractionInPrompt,
        updateEnrollment
      );

      expect(createOrUpdateEnrollment).toHaveBeenCalledWith({
        enrollment: enrollmentToUpdate,
      });
      expect(output).toMatchSnapshot();
    });

    it('displays an error if update fails', async () => {
      createOrUpdateEnrollment.mockRejectedValue("Pas d'update désolé");

      const output = await handleEnrollmentSubmission(
        actionConfiguration,
        setIntendedAction,
        enrollmentToUpdate,
        waitForUserInteractionInPrompt,
        updateEnrollment
      );

      expect(output).toMatchSnapshot();
    });
  });
});
