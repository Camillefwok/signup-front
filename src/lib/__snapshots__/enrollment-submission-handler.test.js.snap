// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When submitting the enrollment form with the destroy action calls the delete endpoint 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": true,
  "successMessages": Array [],
}
`;

exports[`When submitting the enrollment form with the notify action calls for the enrollment state update after user validates prompt 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": true,
  "successMessages": Array [],
}
`;

exports[`When submitting the enrollment form with the notify action waits for user prompt 1`] = `
Object {
  "errorMessages": Array [],
  "redirectToHome": false,
  "successMessages": Array [],
}
`;
